pkg/codegen/nodejs/gen_program_test.go

```diff
@@ -29,41 +30,51 @@ func nodejsCheck(t *testing.T, path string, dependencies codegen.StringSet) {
 	ex, err := executable.FindExecutable("yarn")
 	require.NoError(t, err, "Could not find yarn executable")
 	dir := filepath.Dir(path)
-	pkgs := nodejsPackages(dependencies)
+	pkgs := nodejsPackages(t, dependencies)
 	// We delete and regenerate package files for each run.
-	packageJSON := filepath.Join(dir, "package.json")
-	if err := os.Remove(packageJSON); !os.IsNotExist(err) {
-		require.NoError(t, err, "Failed to delete %s", packageJSON)
+	packageJSONPath := filepath.Join(dir, "package.json")
+	if err := os.Remove(packageJSONPath); !os.IsNotExist(err) {
+		require.NoError(t, err, "Failed to delete %s", packageJSONPath)
 	}
 	yarnLock := filepath.Join(dir, "yarn.lock")
 	if err := os.Remove(yarnLock); !os.IsNotExist(err) {
 		require.NoError(t, err, "Failed to delete %s", yarnLock)
 	}
 
-	err = integration.RunCommand(t, "link @pulumi/pulumi",
-		[]string{ex, "link", "@pulumi/pulumi"},
-		dir, &integration.ProgramTestOptions{})
-	require.NoError(t, err, "Failed to link @pulumi/pulumi")
-	for _, pkg := range pkgs {
-		err = integration.RunCommand(t, "yarn add and install",
-			[]string{ex, "add", pkg}, dir, &integration.ProgramTestOptions{})
-		require.NoError(t, err, "Could not install package: %q", pkg)
+	pkgInfo := npmPackage{
+		Dependencies: map[string]string{
+			"@pulumi/pulumi": "latest",
+		},
+		DevDependencies: map[string]string{
+			"@types/node": "latest",
+			"typescript":  "latest",
+		},
+	}
+	for pkg, v := range pkgs {
+		pkgInfo.Dependencies[pkg] = v
 	}
+	pkgJSON, err := json.MarshalIndent(pkgInfo, "", "    ")
+	require.NoError(t, err)
+	err = os.WriteFile(packageJSONPath, pkgJSON, 0600)
+	require.NoError(t, err)
+
+	err = integration.RunCommand(t, "Install dependencies",
+		[]string{ex, "install"},
+		dir, &integration.ProgramTestOptions{})
+	require.NoError(t, err, "Failed install")
+
 	err = integration.RunCommand(t, "tsc check",
 		[]string{ex, "run", "tsc", "--noEmit", filepath.Base(path)}, dir, &integration.ProgramTestOptions{})
 	require.NoError(t, err, "Failed to build %q", path)
 }
 
 // Returns the nodejs equivalent to the hcl2 package names provided.
-func nodejsPackages(deps codegen.StringSet) []string {
-	if len(deps) == 0 {
-		return []string{"@pulumi/pulumi@3.7.0"}
-	}
-	result := make([]string, len(deps))
-	for i, d := range deps.SortedValues() {
+func nodejsPackages(t *testing.T, deps codegen.StringSet) map[string]string {
+	result := make(map[string]string, len(deps))
+	for _, d := range deps.SortedValues() {
 		r := fmt.Sprintf("@pulumi/%s", d)
 		v := func(s string) {
-			r = fmt.Sprintf("%s@%s", r, s)
+			result[r] = "^" + s
```

Comment:
Could you add variable names or comments to explain what's going on here? `v` and `s` aren't that helpful to me.